### DISCUSSION TITLE
Dispatcher.UIThread.Invoke is not successfully setting a property

### DISCUSSION THREAD

**Discussion by reelthymeos** [AvaloniaUI/Avalonia#16909] [Q&A]
Hi,

I have text that is not being displayed in a TextBlock. The 'Text' property of that block is bound to my property named 'Response' ( Text = "{Binding Response}"), and I structured that property how one usually would:

```
private string response;

public string Response
{
	get
	{
		return response;
	}

	private set
	{
		response = value;
		RaisePropertyChanged(nameof(Response));
	}
}
```
I have a button in my View which has its Command property bound to a Command method. This method called OnSendCommand is shown below. 

```



private void OnSendCommand()
{
	string myString= string.Empty;

	GetMyString(ref myString);
			
	Dispatcher.UIThread.InvokeAsync(new Action(() => { Response = myString; }));
}
```

When I press the button and this command fires, it executes the Dispatcher line to update the property 'Response', but the property continues to show a value of null after the statement is executed. Subsequently when the Getter is called for the property, it returns null and thus the TextBlock is empty. The Command method fires again and updates the property correctly, but by then the Getter has been called already so the text block never updates. 

STEPS I TRIED:

1. Tried to use Dispatcher.UIThread.Post(), this did not work
2. I tried using the Dispatcher.UIThread.InvokeAsync)() with the priority parameter set to DispatcherPriority.MaxValue but this also didn't help.
3. Also tried plain-old Invoke() and that didn't help

EDIT:

I have attached a project that is a minimum sample that reproduces this issue.

**Comment by IanRawley** [comment]
Can you show the XAML for the TextBlock, or how the TextBlock is otherwise being bound to Response?  Also does GetMyString actually get a non-empty string?

**Reply by reelthymeos** [reply]
Here is the XAML for the text block:

`<TextBlock Grid.Row="4" Grid.ColumnSpan="4" Width="600" Height="250" TextWrapping="WrapWithOverflow" Text="{Binding Response}"></TextBlock>`

GetMyString returns a non-empty string.

**Comment by reelthymeos** [comment]
Here is a minimum sample that reproduces the issue: https://github.com/reelthymeos/MinimumSample4563

**Reply by timunie** [reply]
Thx

**Reply by reelthymeos** [reply]
Hi, has anyone had a chance to look into this further? Thanks!

**Reply by stevemonaco** [reply]
There are a lot of issues:

1. Your repro needs a .gitignore so you don't commit binaries.
2. It has no reference to $(AvaloniaVersion), so I'm not sure what version you're using. I switched to 11.1.3.
3. Your project has 3 `PropertyChanged` notifications between Mvvm Toolkit, ReactiveUI, and your own bespoke implementation. This is your core problem: remove your `PropertyChanged` stuff instead of hiding ReactiveUI's implementation because you're deriving `MainViewModel` from `ReactiveObject`. Just use `this.RaiseAndSetIfChanged(ref response, value);`. Pick one and stick with it, don't try to mix-and-match as a beginner.
4. Your string building for Response is sending in a function instead of calling it.

Probably some other things, too. But it works fine after addressing those points.

**Reply by reelthymeos** [reply]
Thanks @stevemonaco . Can you elaborate on your last point? 

> Your string building for Response is sending in a function instead of calling it.

I'm not sure what that is referring to.

**Reply by stevemonaco** [reply]
```cs
Response += $"{DateTime.Now.ToShortDateString}\t\t\t\t\t\t{Command}\t\t\t\t\t\t{thisResponse}";
```

You should use `ToShortDateString()`.

**Reply by reelthymeos** [reply]
Regarding the property changed notifications, I believe the MVVM Toolkit reference is in my project because when I was setting up the initial Avalonia project I chose that as the framework. If I remove that reference from my project entirely, do you think it will cause any issues? I am not actually inheriting from ViewModelBase for any of my VMs. I did that custom implementation of RaisePropertyChanged because that is the example I got from the Avalonia tutorial on MVVM. 

So in summary, some of my VMs implement the INofifyPropertyChanged interface, and the others are ReactiveObjects. Is there an issue with having such a mix in the same project? I'm talking about my sub-VMs here, not the main VM (which will be comprised of all those sub-VMs).

**Reply by stevemonaco** [reply]
There's no fundamental issue with mixing if you know what you're doing. It's even somewhat common though I avoid it. Beginners don't quite have a grasp on what's happening which leads to difficult troubleshooting, so I don't advise mixing. If you're using a ViewLocator, make sure it matches on `INotifyPropertyChanged` instead of only `ReactiveObject`. You won't have the RxUI stuff inside of a plain `INotifyPropertyChanged` implementation either.
